reasoner/planner/AnswerCountEstimator.java

```diff
@@ -685,27 +685,36 @@ private double estimateInferredAnswerCount(Concludable concludable, Set<Variable
-                            Set<Variable> inferredRolePlayers = iterate(ruleSideVariables).filter(Variable::isThing)
+                            Set<Variable> inferredRolePlayers = iterate(recursivePlayers).filter(Variable::isThing)
                                     .filter(v -> iterate(v.inferredTypes()).anyMatch(type -> answerCountEstimator.logicMgr.rulesConcluding(type).hasNext()))
                                     .toSet();
 
                             iterate(inferredRolePlayers)
                                 .forEachRemaining(inferrableRoleplayer -> {
-                                    if (typeBasedUpperBoundsPerPlayer.getOrDefault(inferrableRoleplayer, 1.0) < inferredEstimateForThisBranch) {
-                                        typeBasedUpperBoundsPerPlayer.merge(inferrableRoleplayer, 1.0, (x,y) -> 1.0);
+                                    if (typeBasedEstimateForRecursivePlayer.getOrDefault(inferrableRoleplayer, 1.0) < inferredEstimateForThisBranch) {
+                                        typeBasedEstimateForRecursivePlayer.merge(inferrableRoleplayer, 1.0, (x,y) -> 1.0);
                                     }
                                 });
 
-                            double replacementForInferred = typeBasedUpperBoundsPerPlayer.size() < ruleSideVariables.size() ? inferredEstimateForThisBranch : 1.0;
+                            double replacementForInferred = !inferredRolePlayers.isEmpty() ? inferredEstimateForThisBranch : 1.0;
 
-                            // All possible combinations assumes full-connectivity, producing an unrealistically large/worst-case estimate.
+                            // All possible combinations (of non-recursive players) assumes full-connectivity, producing an unrealistically large/worst-case estimate.
                             // We take the square root of all possible combinations as a more realistic estimate.
-                            double heuristicUpperBound = Math.ceil(Math.sqrt(iterate(typeBasedUpperBoundsPerPlayer.values()).reduce(replacementForInferred, (x,y) -> x * y)));
+                            Set<Variable> nonRecursivePlayers = iterate(ruleSideVariables).filter(v -> !recursivePlayers.contains(v)).toSet();
+                            double estimateForNonRecursivePlayers = answerCountEstimator.estimateAnswers(conditionBranch.conjunction(), nonRecursivePlayers);
+                            double rootOfAll = Math.max(1.0, Math.sqrt(iterate(typeBasedEstimateForRecursivePlayer.values()).reduce(replacementForInferred, (x,y) -> x * y)));
+                            double recursivelyConnectedEstimate = iterate(typeBasedEstimateForRecursivePlayer.values()).map(x -> rootOfAll * Math.max(1.0, Math.sqrt(x))).reduce(0.0, Double::sum);
+                            double heuristicUpperBound = Math.ceil(estimateForNonRecursivePlayers * recursivelyConnectedEstimate);
```

Comment:
The initial `sqrt(n)` estimate did not make sense, as it inferred a linear number of inferred relations for transitivity - what we intended was for a single node to be connected to at most sqrt(n) other nodes.

This version estimates O(n^(3/2)) inferred relations, yielding the average connectivity of a single as sqrt(n)